File: word-ladder.cc
using pii = pair<int, int>;

class Solution {
  bool connected(string& a, string& b) {
    int diff = 0;
    for (int i = 0; i < a.size(); i++) {
      diff += a[i] != b[i];
    }
    return diff == 1;
  }

  vector<int> graph[5001];
  bool vis[5001];

public:
  int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
    int n = wordList.size(), ei = -1;

    queue<pii> q;
    memset(vis, 0, sizeof(vis));

    for (int i = 0; i < n; i++) {
      if (wordList[i] == endWord) ei = i;
      if (connected(wordList[i], beginWord)) {
        q.push({i, 1});
        vis[i] = true;
      }
    }

    for (int i = 0; i < n; i++) for (int j = i + 1; j < n; j++) {
      if (connected(wordList[i], wordList[j])) {
        graph[i].push_back(j);
        graph[j].push_back(i);
      }
    }

    while (!q.empty()) {
      auto [v, d] = q.front();
      q.pop();

      if (v == ei) return d + 1;

      for (int u : graph[v]) {
        if (!vis[u]) {
          vis[u] = true;
          q.push({u, d + 1});
        }
      }
    }

    return 0;
  }
};

File: word-ladder.rb
# @param {String} begin_word
# @param {String} end_word
# @param {String[]} word_list
# @return {Integer}
def ladder_length(begin_word, end_word, word_list)
  word_length = begin_word.size
  graph = Hash.new {|h, k| h[k] = []}
  
  (word_list + [begin_word]).each do |word|
    candidate_words = word_length.times.map do |i|
      word[0...i] + '*' + word[i + 1..-1]
    end
    candidate_words.each {|cw| graph[cw] << word}
  end
  visited = (word_list + [begin_word]).map {|w| [w, false]}.to_h

  result = 0
  queue = []
  queue << [begin_word, 1]
  visited[begin_word] = true
  until queue.empty?
    item, idx = queue.shift
    if item == end_word
      result = idx
      break
    end

    visited[item] = true
    word_length.times do |i|
      candidate_word = item[0...i] + '*' + item[i + 1..-1]
      graph[candidate_word].each {|dst| queue << [dst, idx + 1] unless visited[dst]}
    end
  end

  result
end
